backend/generalhw.pm

```diff
@@ -27,39 +29,40 @@ sub get_cmd ($self, $cmd) {
     my $dir = get_required_var('GENERAL_HW_CMD_DIR');
     die 'GENERAL_HW_CMD_DIR is not pointing to a directory' unless -d $dir;
 
-    my %GENERAL_HW_ARG_VARIABLES_BY_CMD = ('GENERAL_HW_FLASH_CMD' => 'GENERAL_HW_FLASH_ARGS', 'GENERAL_HW_SOL_CMD' => 'GENERAL_HW_SOL_ARGS', 'GENERAL_HW_POWERON_CMD' => 'GENERAL_HW_POWERON_ARGS', 'GENERAL_HW_POWEROFF_CMD' => 'GENERAL_HW_POWEROFF_ARGS');
+    my %GENERAL_HW_ARG_VARIABLES_BY_CMD = (
+        'GENERAL_HW_FLASH_CMD' => 'GENERAL_HW_FLASH_ARGS',
+        'GENERAL_HW_SOL_CMD' => 'GENERAL_HW_SOL_ARGS',
+        'GENERAL_HW_POWERON_CMD' => 'GENERAL_HW_POWERON_ARGS',
+        'GENERAL_HW_POWEROFF_CMD' => 'GENERAL_HW_POWEROFF_ARGS',
+        'GENERAL_HW_STORE_ASSET_CMD' => 'GENERAL_HW_STORE_ASSET_ARGS',
+    );
     my $args = get_var($GENERAL_HW_ARG_VARIABLES_BY_CMD{$cmd}) if get_var($GENERAL_HW_ARG_VARIABLES_BY_CMD{$cmd});
 
-    # Append HDD infos to flash script
-    if ($cmd eq 'GENERAL_HW_FLASH_CMD' and get_var('HDD_1')) {
-        my $numdisks = get_var('NUMDISKS') // 1;
-        for my $i (1 .. $numdisks) {
-            # Pass path of HDD
-            $args .= " " . get_required_var("HDD_$i");
-            # Pass size of HDD
-            my $size = get_var("HDDSIZEGB_$i");
-            $size //= get_var('HDDSIZEGB') // 10;
-            $args .= " $size" . 'G';
-        }
-    }
-
     $cmd = get_required_var($cmd);
     $cmd = "$dir/" . basename($cmd);
     $cmd .= " $args" if $args;
     return $cmd;
 }
 
-sub run_cmd ($self, @args) {
-    my @full_cmd = split / /, $self->get_cmd($args[0]);
+sub run_cmd ($self, $cmd, @extra_args) {
+    my @full_cmd = split / /, $self->get_cmd($cmd);
+
+    push @full_cmd, @extra_args;
 
     my ($stdin, $stdout, $stderr, $ret);
-    eval { $ret = IPC::Run::run([@full_cmd], \$stdin, \$stdout, \$stderr) };
-    die "Unable to run command '@full_cmd' (deduced from test variable $args[0]): $@\n" if $@;
+
+    # Do not let the SIGCHLD handler of Mojo::IOLoop steal the exit code from IPC::Run
+    session->collect_status(0);
+    {
+        my $guard = scope_guard sub { session->collect_status(1); };
```

Comment:
Good use of a scope guard. I would probably just replace the two calls of `collect_status` with a single `local $SIG{CHLD};` here though.

Reply:
I'm a bit worried of consequences of RWP-started processes, if any exits at this very time. Wouldn't that be a problem? If not, then sure, I can change the SIGCHLD handler.

Reply:
Would you like me to change it to `local $SIG{CHLD}`?

Reply:
Both should be fine here. But you should change the comment, since `Mojo::IOLoop` doesn't have a `SIGCHLD` handler (it is RWP only).

Reply:
It is true that if an RWP-started process exists at this very time it will fail to emit the collected event for that process. However, we only lose a log message so I suppose we can live with that. And yes, this is RWP-specific (and nothing `Mojo::IOLoop` does in general).